[Ruby] new concept exercise - `instance-variables` (#2023)
exercism/v3: languages/ruby/exercises/concept/instance-variables/.docs/introduction.md

> ## Object state, instance variables

Reviewer: I really like the introduction!

> Objects can hold their own state by setting _instance variables_. _Instance variables_ are created a prefixing `@` to the variable name.

Reviewer 2: ```suggestion
Objects can hold their own state by setting _instance variables_. _Instance variables_ are created by prefixing `@` to the variable name.
```

> Objects can hold their own state by setting _instance variables_. _Instance variables_ are created by prefixing `@` to the variable name.

Reviewer 3: ```suggestion
Objects can hold their own state by setting _instance variables_., which are created by prefixing `@` to a variable name.
```

```ruby

> @instance_variable = 2

Reviewer 3: I suggest not actually calling this `instance_variable` in case people misconstrue that as a formal part of the naming.

```suggestion
@name = 2
```

```

> This is often done by defining an `initialize` method to set an initial state.

Reviewer 3: ```suggestion
Instance variables are often given initial state in an `initialize` method, which is automatically called when a class is instantiated.
```

> Instance variables are often given initial state in an `initialize` method, which is automatically called when a class is instantiated.

Reviewer 2: ```suggestion
Objects usually set their initial state in an `initialize` method, which is automatically called when calling `new` on a class.
```

```ruby
class Airplane
  def initialize
    @wings = 2
  end
end
```

The `initialize` method may also take arguments, so that each instance can start with a custom state:

```ruby
class Suitcase
  def initialize(locked)
    @locked = locked
  end
end
```

> Consider _instance variables_ to be private from external read and writes. _Class instance methods_ should be used for getting and setting instance variables:

Reviewer 2: ```suggestion
Consider _instance variables_ to be private from external read and writes. _Instance methods_ should be used for getting and setting instance variables:
```

Class instance methods should not be used, generally.  We can introduce those later.

Author: I think I just meant instance methods anyway.

> Consider _instance variables_ to be private from external read and writes. _Instance methods_ should be used for getting and setting instance variables:

Reviewer 2: ```suggestion
Consider _instance_ variables to be private from external read and writes. _Instance_ methods should be used for getting and setting instance variables:
```

```ruby
class Suitcase
  #...

>   def locked? # Methods returning true or false should have trailing `?`

Reviewer 2: Methods that can be used as a Boolean often have a trailing `?` as part of the name.  We call these "query methods".

This hints at the true and false, while leaving it open.  The realization of why it is stated this way becomes apparent when we talk about "truthy" and "falsey".

Author: Awesome, I didn't know that they were called that, will integrate the change!

Author: ```suggestion
  def locked? # Query methods should be named with a trailing `?`
```

    @locked
  end

>   def unlock! # Methods which mutate state should have trailing `!`

Reviewer 2: Unless a method name is obvious that it mutates state, the last character of a method that changes state will have a `!` as the last character.

For example:  A method named `delete` does not need the `!` character, but a method named `color!` would indicate that the color would change.

Reviewer 2: This was resolved, but Ruby convention is that the bang is unnecessary to state mutation, unless it is unclear.  Unlocking something is clearly a mutating behavior.  You can not, for example, unlock a door without changing its state, it should be intuitive that unlocking a pass a pass would mutate the pass.

    @locked = false
  end
end
```

## Nil

> [Nil][nil-dictionary] is an English word meaning "nothing" or "zero". In Ruby, `nil` is a value used to express the _absence_ of an object. In other programming languages, `null` or `none` values may play a similar role.

Reviewer 2: This is tricky, because `nil` itself is an object, it is not a "value" per se.  I mean, in that all objects are values, but it is an object used to express "nothingness".

Author: ```suggestion
[Nil][nil-dictionary] is an English word meaning "nothing" or "zero". In Ruby, `nil` is used to express the _absence_ of an object. In other programming languages, `null` or `none` values may play a similar role.
```

```ruby
# I do not have a favorite color
favorite_color = nil
```

> Ruby gives any undefined instance variable the default value of `nil` when it is accessed before being set.

Reviewer 2: Ruby will provide a `nil` value to any instance method upon first encounter, unless it is explicitly assigned.

Author: Do you mean variable or method?

Reviewer 2: I do mean variable.

Author: ```suggestion
Ruby gives any instance variable the default value of `nil` when it is first encountered, until it is set otherwise.
```
